File: Misc/98_lc_2305.java
package Misc;

/*
 * 2305. Fair Distribution of Cookies
 */
class Solution {
    private int res = Integer.MAX_VALUE;

    private void rec(int startingIndex, int[] cookies, int k, int[] possibleDistribution, int cookiesLen) {
        if (startingIndex == cookiesLen) {
            int maxVal = 0;
            for (int val: possibleDistribution) {
                maxVal = Math.max(maxVal, val);
            }
            res = Math.min(res, maxVal);
            return;
        }

        for (int ind=0; ind<k; ind++) {
            if (possibleDistribution[ind] + cookies[startingIndex] < res){
                possibleDistribution[ind] += cookies[startingIndex];
                rec(startingIndex+1, cookies, k, possibleDistribution, cookiesLen);
                possibleDistribution[ind] -= cookies[startingIndex];
            } 
        }
    }

    public int distributeCookies(int[] cookies, int k) {
        int cookiesLen = cookies.length;
        rec(0, cookies, k, new int[k], cookiesLen);
        return res;
    }
}

File: Misc/98_lc_2305.py
'''
2305. Fair Distribution of Cookies
'''
from typing import List

class Solution:
    def distributeCookies(self, cookies: List[int], k: int) -> int:
        arr = [0 for _ in range(k)]
        n = len(cookies)
        res = 1_000_000_000
        
        def rec(i):
            nonlocal res
            if i == n:
                arrMax = max(arr)
                res = min(res, arrMax)
                return
            
            for ind in range(k):
                if arr[ind] + cookies[i] < res:
                    arr[ind] += cookies[i]
                    rec(i+1)
                    arr[ind] -= cookies[i]

        rec(0)
        return res
